<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Application of poisson mash and mash to single cell cytokines data (B cells)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">poisson_mash_random_effects</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/poisson_mash_random_effects">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Application of poisson mash and mash to single cell cytokines data (B cells)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2020-11-5</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-11-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>poisson_mash_random_effects/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201102code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201102)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201102code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201102)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstreef62055315ae7ab998fb222ed6c49714fde7a5105targetblankf620553a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/f62055315ae7ab998fb222ed6c49714fde7a5105" target="_blank">f620553</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstreef62055315ae7ab998fb222ed6c49714fde7a5105targetblankf620553a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/f62055315ae7ab998fb222ed6c49714fde7a5105" target="_blank">f620553</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    output/.DS_Store
    Ignored:    output/Ly6C-/.DS_Store

Untracked files:
    Untracked:  output/B_cells/
    Untracked:  output/Ly6C-/condition_colors.Rds
    Untracked:  output/Ly6C-/sim_null/ed_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_fit_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_posterior_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/raw_data3.Rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/application_B_cells.Rmd</code>) and HTML (<code>docs/application_B_cells.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/f62055315ae7ab998fb222ed6c49714fde7a5105/analysis/application_B_cells.Rmd" target="_blank">f620553</a>
</td>
<td>
yushaliu
</td>
<td>
2020-11-05
</td>
<td>
upload constrained estimation of rank-1 covariances
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>We apply poisson mash and mash to a subset of the single cell cytokines dataset, which include gene expresion data of <span class="math inline">\(15,705\)</span> B cells from <span class="math inline">\(R=25\)</span> conditions (i.e., different cytokine treatments) in the second batch. We apply both methods to the <span class="math inline">\(9,968\)</span> genes with at least <span class="math inline">\(50\)</span> reads across all cells.</p>
<pre class="r"><code>cols.trt &lt;- readRDS(&quot;output/B_cells/condition_colors.Rds&quot;)
trts &lt;- names(cols.trt)</code></pre>
</div>
<div id="poisson-mash" class="section level2">
<h2>Poisson mash</h2>
<p>We apply poisson mash to the <span class="math inline">\(9,968 \times 25\)</span> count data matrix <span class="math inline">\(X\)</span>, where <span class="math inline">\(X_{jr}\)</span> represents the total counts of gene <span class="math inline">\(j\)</span> in all cells from condition <span class="math inline">\(r\)</span>. To estimate the data-driven prior covariances <span class="math inline">\(U_k\)</span>, we fit the poisson mash model without scaling parameters <span class="math inline">\(w_l\)</span> to a subset of genes (<span class="math inline">\(\tilde{J} = 1,964\)</span>) that are differentially expressed across conditions, which are identified based on a conditional multinomial goodness-of-fit test. Two versions of ED step are fitted, depending on whether we restrict the vectors <span class="math inline">\(u_g\)</span> forming the rank-1 data-driven prior covariances <span class="math inline">\(U_g = u_g u_g&#39;\)</span> to have the same sign across conditions or not.</p>
<div id="compare-the-elbos-in-the-ed-step-constrained-vs-unconstrained" class="section level3">
<h3>Compare the ELBOs in the ED step (constrained vs unconstrained)</h3>
<pre class="r"><code>## load in the ED results
pois.ed &lt;- readRDS(&quot;output/B_cells/pois_ed_orig1.Rds&quot;)
pois.ed.v2 &lt;- readRDS(&quot;output/B_cells/pois_ed_orig1_constrained.Rds&quot;)

## running time per iteration in ED
pois.ed$runtime/length(pois.ed$ELBO)</code></pre>
<pre><code>     user    system   elapsed 
18.803146  0.026703 18.830324 </code></pre>
<pre class="r"><code>pois.ed.v2$runtime/length(pois.ed.v2$ELBO)</code></pre>
<pre><code>     user    system   elapsed 
19.468241  0.029906 19.508591 </code></pre>
<pre class="r"><code>## look at the ELBO
plot(pois.ed$ELBO, xlab = &quot;niter&quot;, ylab = &quot;elbo&quot;, main=&quot;Fitting ED step in Poisson mash&quot;)
lines(pois.ed.v2$ELBO, col=&quot;red&quot;, lty=4)
legend(&quot;bottomright&quot;, legend=c(&quot;unconstrained&quot;, &quot;constrained&quot;), text.col=c(&quot;black&quot;, &quot;red&quot;), 
       col=c(&quot;black&quot;, &quot;red&quot;), lty=c(1,4), bty=&quot;n&quot;)</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="look-at-the-elbos-in-the-model-fitting-step-constrained-vs-unconstrained" class="section level3">
<h3>Look at the ELBOs in the model fitting step (constrained vs unconstrained)</h3>
<pre class="r"><code>## load in the poisson mash fit results
pois.res &lt;- readRDS(&quot;output/B_cells/pois_mash_fit_orig1.Rds&quot;)
pois.res.v2 &lt;- readRDS(&quot;output/B_cells/pois_mash_fit_orig1_constrained.Rds&quot;)

## running time for fitting poisson mash to all genes
pois.res$runtime</code></pre>
<pre><code>     user    system   elapsed 
27470.142    88.328 27556.692 </code></pre>
<pre class="r"><code>pois.res.v2$runtime</code></pre>
<pre><code>     user    system   elapsed 
29903.938   149.299 30069.982 </code></pre>
<pre class="r"><code>## look at ELBO
plot(pois.res$ELBO, xlab = &quot;niter&quot;, ylab = &quot;elbo&quot;, main=&quot;Fitting Poisson mash to all genes&quot;)
lines(pois.res.v2$ELBO, col=&quot;red&quot;, lty=4)
legend(&quot;bottomright&quot;, legend=c(&quot;unconstrained&quot;, &quot;constrained&quot;), text.col=c(&quot;black&quot;, &quot;red&quot;), 
       col=c(&quot;black&quot;, &quot;red&quot;), lty=c(1,4), bty=&quot;n&quot;)</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## compute the weights of different prior covariances
wlist &lt;- c(1e-4, 1e-3, 2.5e-3, 5e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32, 0.64, 1, 2, 4, 8, 16, 25, 36)

pi.mat &lt;- matrix(pois.res$pi, ncol=length(wlist), byrow=TRUE)
rownames(pi.mat) &lt;- c(names(pois.ed$Ulist), paste0(&quot;e_&quot;, 1:25), names(pois.ed$ulist))
colnames(pi.mat) &lt;- paste0(&quot;w=&quot;, wlist)

pi.mat.v2 &lt;- matrix(pois.res.v2$pi, ncol=length(wlist), byrow=TRUE)
rownames(pi.mat.v2) &lt;- c(names(pois.ed.v2$Ulist), paste0(&quot;e_&quot;, 1:25), names(pois.ed.v2$ulist))
colnames(pi.mat.v2) &lt;- paste0(&quot;w=&quot;, wlist)</code></pre>
</div>
<div id="estimate-of-data-driven-covariances-unconstrained" class="section level3">
<h3>Estimate of data-driven covariances (unconstrained)</h3>
<pre class="r"><code>### Estimate of eigenvectors of the data-driven covariance initialized by rank-5 PCA
eig.tPCA &lt;- eigen(pois.ed$Ulist[[1]])
pve.tPCA &lt;- eig.tPCA$values/sum(eig.tPCA$values)

par(mfrow=c(2,2))

for (k in 1:4){
  v &lt;- eig.tPCA$vectors[,k]
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(pois.ed$Ulist)[1], &quot;: weight &quot;, round(rowSums(pi.mat)[1],3), 
                                 &quot;:\nEigenvector &quot;, k, &quot; (pve = &quot;, round(pve.tPCA[k],3), &quot;)&quot;))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### Estimate of eigenvectors of rank-1 data-driven covariance matrices
for (k in 1:4){
  v &lt;- as.numeric(pois.ed$ulist[[k]])
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(pois.ed$ulist)[k], &quot;: weight &quot;, round(rowSums(pi.mat)[27+k],3)))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="estimate-of-data-driven-covariances-constrained" class="section level3">
<h3>Estimate of data-driven covariances (constrained)</h3>
<pre class="r"><code>### Estimate of eigenvectors of the data-driven covariance initialized by rank-5 PCA
eig.tPCA &lt;- eigen(pois.ed.v2$Ulist[[1]])
pve.tPCA &lt;- eig.tPCA$values/sum(eig.tPCA$values)

par(mfrow=c(2,2))

for (k in 1:4){
  v &lt;- eig.tPCA$vectors[,k]
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(pois.ed.v2$Ulist)[1], &quot;: weight &quot;, round(rowSums(pi.mat.v2)[1],3), 
                                 &quot;:\nEigenvector &quot;, k, &quot; (pve = &quot;, round(pve.tPCA[k],3), &quot;)&quot;))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### Estimate of eigenvectors of rank-1 data-driven covariance matrices
for (k in 1:4){
  v &lt;- as.numeric(pois.ed.v2$ulist[[k]])
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(pois.ed.v2$ulist)[k], &quot;: weight &quot;, round(rowSums(pi.mat.v2)[27+k],3)))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="mash-with-random-effects" class="section level2">
<h2>Mash with random effects</h2>
<p>We then fit mash with gene-specific means and random effects to the <span class="math inline">\(9,968 \times 25\)</span> matrix of point estimates and standard errors, which are calculated from the normalized single cell data. To estimate the data-driven prior covariances <span class="math inline">\(U_k\)</span>, we fit mash without scaling parameters <span class="math inline">\(w_l\)</span> to a subset of genes (<span class="math inline">\(\tilde{J} = 1,544\)</span>) that are differentially expressed across conditions, which are defined here as genes for which <span class="math inline">\(\max_r |Z_{jr}| \geq 3.5\)</span>. Two versions of ED step are fitted, depending on whether we restrict the vectors <span class="math inline">\(u_g\)</span> forming the rank-1 data-driven prior covariances <span class="math inline">\(U_g = u_g u_g&#39;\)</span> to have the same sign across conditions or not.</p>
<div id="compare-the-log-likelihood-in-the-ed-step-constrained-vs-unconstrained" class="section level3">
<h3>Compare the log likelihood in the ED step (constrained vs unconstrained)</h3>
<pre class="r"><code>## load in the ED results
mash.ed &lt;- readRDS(&quot;output/B_cells/mash_ed_orig1.Rds&quot;)
mash.ed.v2 &lt;- readRDS(&quot;output/B_cells/mash_ed_orig1_constrained.Rds&quot;)

## running time per iteration in ED
mash.ed$runtime/length(mash.ed$llik)</code></pre>
<pre><code>    user   system  elapsed 
7.896116 0.012108 7.910904 </code></pre>
<pre class="r"><code>mash.ed.v2$runtime/length(mash.ed.v2$llik)</code></pre>
<pre><code>    user   system  elapsed 
9.460506 0.017562 9.485494 </code></pre>
<pre class="r"><code>## look at the log likelihood
plot(mash.ed$llik, xlab = &quot;niter&quot;, ylab = &quot;log likelihood&quot;, main=&quot;Fitting ED step in mash with random effects&quot;)
lines(mash.ed.v2$llik, col=&quot;red&quot;, lty=4)
legend(&quot;bottomright&quot;, legend=c(&quot;unconstrained&quot;, &quot;constrained&quot;), text.col=c(&quot;black&quot;, &quot;red&quot;), 
       col=c(&quot;black&quot;, &quot;red&quot;), lty=c(1,4), bty=&quot;n&quot;)</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="look-at-the-log-likelihood-in-the-model-fitting-step-constrained-vs-unconstrained" class="section level3">
<h3>Look at the log likelihood in the model fitting step (constrained vs unconstrained)</h3>
<pre class="r"><code>## load in the mash fit results
mash.res &lt;- readRDS(&quot;output/B_cells/mash_fit_orig1.Rds&quot;)
mash.res.v2 &lt;- readRDS(&quot;output/B_cells/mash_fit_orig1_constrained.Rds&quot;)

## running time for fitting poisson mash to all genes
mash.res$runtime</code></pre>
<pre><code>     user    system   elapsed 
10975.746    83.268 11057.428 </code></pre>
<pre class="r"><code>mash.res.v2$runtime</code></pre>
<pre><code>     user    system   elapsed 
12981.151   114.546 13106.680 </code></pre>
<pre class="r"><code>## look at the log likelihood
plot(mash.res$llik, xlab = &quot;niter&quot;, ylab = &quot;log likelihood&quot;, main=&quot;Fitting mash with random effects to all genes&quot;)
lines(mash.res.v2$llik, col=&quot;red&quot;, lty=4)
legend(&quot;bottomright&quot;, legend=c(&quot;unconstrained&quot;, &quot;constrained&quot;), text.col=c(&quot;black&quot;, &quot;red&quot;), 
       col=c(&quot;black&quot;, &quot;red&quot;), lty=c(1,4), bty=&quot;n&quot;)</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## compute the weights of different prior covariances
wlist &lt;- c(1e-6, 4e-6, 1e-5, 4e-5, 1e-4, 2.5e-4, 5e-4, 1e-3, 2.5e-3, 5e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32)

pi.mat &lt;- matrix(mash.res$pi, ncol=length(wlist), byrow=TRUE)
rownames(pi.mat) &lt;- c(names(mash.ed$Ulist), paste0(&quot;e_&quot;, 1:25), names(mash.ed$ulist))
colnames(pi.mat) &lt;- paste0(&quot;w=&quot;, wlist)

pi.mat.v2 &lt;- matrix(mash.res.v2$pi, ncol=length(wlist), byrow=TRUE)
rownames(pi.mat.v2) &lt;- c(names(mash.ed.v2$Ulist), paste0(&quot;e_&quot;, 1:25), names(mash.ed.v2$ulist))
colnames(pi.mat.v2) &lt;- paste0(&quot;w=&quot;, wlist)</code></pre>
</div>
<div id="estimate-of-data-driven-covariances-unconstrained-1" class="section level3">
<h3>Estimate of data-driven covariances (unconstrained)</h3>
<pre class="r"><code>### Estimate of eigenvectors of the data-driven covariance initialized by rank-5 PCA
eig.tPCA &lt;- eigen(mash.ed$Ulist[[1]])
pve.tPCA &lt;- eig.tPCA$values/sum(eig.tPCA$values)

par(mfrow=c(2,2))

for (k in 1:4){
  v &lt;- eig.tPCA$vectors[,k]
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(mash.ed$Ulist)[1], &quot;: weight &quot;, round(rowSums(pi.mat)[1],3), 
                                 &quot;:\nEigenvector &quot;, k, &quot; (pve = &quot;, round(pve.tPCA[k],3), &quot;)&quot;))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### Estimate of eigenvectors of rank-1 data-driven covariance matrices
for (k in 1:4){
  v &lt;- as.numeric(mash.ed$ulist[[k]])
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(mash.ed$ulist)[k], &quot;: weight &quot;, round(rowSums(pi.mat)[27+k],3)))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-8-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="estimate-of-data-driven-covariances-constrained-1" class="section level3">
<h3>Estimate of data-driven covariances (constrained)</h3>
<pre class="r"><code>### Estimate of eigenvectors of the data-driven covariance initialized by rank-5 PCA
eig.tPCA &lt;- eigen(mash.ed.v2$Ulist[[1]])
pve.tPCA &lt;- eig.tPCA$values/sum(eig.tPCA$values)

par(mfrow=c(2,2))

for (k in 1:4){
  v &lt;- eig.tPCA$vectors[,k]
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(mash.ed.v2$Ulist)[1], &quot;: weight &quot;, round(rowSums(pi.mat.v2)[1],3), 
                                 &quot;:\nEigenvector &quot;, k, &quot; (pve = &quot;, round(pve.tPCA[k],3), &quot;)&quot;))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### Estimate of eigenvectors of rank-1 data-driven covariance matrices
for (k in 1:4){
  v &lt;- as.numeric(mash.ed.v2$ulist[[k]])
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
          main = paste0(names(mash.ed.v2$ulist)[k], &quot;: weight &quot;, round(rowSums(pi.mat.v2)[27+k],3)))
}</code></pre>
<p><img src="figure/application_B_cells.Rmd/unnamed-chunk-9-2.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4.6    rprojroot_1.3-2 digest_0.6.25   later_1.0.0    
 [5] R6_2.4.1        backports_1.1.7 git2r_0.27.1    magrittr_1.5   
 [9] evaluate_0.14   stringi_1.4.6   rlang_0.4.7     fs_1.4.1       
[13] promises_1.1.0  whisker_0.4     rmarkdown_2.3   tools_3.6.2    
[17] stringr_1.4.0   glue_1.4.1      httpuv_1.5.2    xfun_0.17      
[21] yaml_2.2.1      compiler_3.6.2  htmltools_0.4.0 knitr_1.30     </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
