<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Explorations of false positives found by poisson mash in simulated single cell data under the null</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">poisson_mash_random_effects</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/poisson_mash_random_effects">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Explorations of false positives found by poisson mash in simulated single cell data under the null</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2020-11-1</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-11-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>poisson_mash_random_effects/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201102code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201102)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201102code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201102)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstree3cd5c87e34b595b91f962e3894c2c46ca224da18targetblank3cd5c87a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/3cd5c87e34b595b91f962e3894c2c46ca224da18" target="_blank">3cd5c87</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstree3cd5c87e34b595b91f962e3894c2c46ca224da18targetblank3cd5c87a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/3cd5c87e34b595b91f962e3894c2c46ca224da18" target="_blank">3cd5c87</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/Ly6C-/.DS_Store

Untracked files:
    Untracked:  output/B_cells/
    Untracked:  output/Ly6C-/condition_colors.Rds
    Untracked:  output/Ly6C-/sim_null/ed_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_fit_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_posterior_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/raw_data3.Rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/sim_null_Ly6C.Rmd</code>) and HTML (<code>docs/sim_null_Ly6C.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/3cd5c87e34b595b91f962e3894c2c46ca224da18/analysis/sim_null_Ly6C.Rmd" target="_blank">3cd5c87</a>
</td>
<td>
yushaliu
</td>
<td>
2020-11-02
</td>
<td>
Add analysis results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-procedure" class="section level2">
<h2>Simulation procedure</h2>
<p>We apply poisson mash and mash to single cell data simulated under the null. More specifically, we start with a subset of the single cell cytokines data from Ly6C- monocytes, which include 1) 1,879 cells from 25 conditions from the second batch, and 2) 6,794 genes that have at least 100 total counts over these 25 conditions. Then we randomly permute the cells with respect to the condition labels so that there no longer exists association between gene expression levels and conditions.</p>
<p>We consider two different ways to simulate null data based on the real dataset:</p>
<p>1). apply the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset;</p>
<p>2). apply independent and different permutation to each gene, so the correlation among genes is no longer preserved on the permuted dataseat.</p>
<p>For each of the scenarios described above, we simulate 5 single cell datasets.</p>
<p>For scenario 1), the number of false positives are <span class="math inline">\(98, 78, 160, 92, 160\)</span> for poisson mash, and <span class="math inline">\(2, 2, 2, 2, 4\)</span> for fixed effects-based mash.</p>
<p>For scenario 2), the number of false positives are <span class="math inline">\(11, 12, 14, 12, 16\)</span> for poisson mash, and <span class="math inline">\(3, 2, 4, 2, 0\)</span> for fixed effects-based mash.</p>
</div>
<div id="exploration-of-fps-found-by-poisson-mash-in-scenario-1" class="section level2">
<h2>Exploration of FPs found by poisson mash in scenario 1)</h2>
<pre class="r"><code>library(Matrix)
library(pheatmap)
library(gridExtra)

### for one replicate dataset, load in the raw single cell count data and the poisson mash fit
data &lt;- readRDS(&quot;output/Ly6C-/sim_null/raw_data3.Rds&quot;)
fit &lt;- readRDS(&quot;output/Ly6C-/sim_null/ed_rep_3.Rds&quot;)
res &lt;- readRDS(&quot;output/Ly6C-/sim_null/pois_mash_fit_rep_3.Rds&quot;)
post &lt;- readRDS(&quot;output/Ly6C-/sim_null/pois_mash_posterior_rep_3.Rds&quot;)

### look at the posterior weights of the genes identified (which are all false positives) 
genes &lt;- which(apply(post$lfsr, 1, min) &lt; 0.05)
wlist &lt;- c(1e-3, 2.5e-3, 5e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32, 0.64, 1, 2, 4, 8, 16, 25, 36, 50)
Ulist &lt;- c(names(fit$Ulist), paste0(&quot;e_&quot;, 1:25), names(fit$ulist))
zeta.max &lt;- rep(NA, length(genes))

for(j in 1:length(genes)){
  zeta &lt;- matrix(res$Ez[genes[j],], ncol=length(wlist), byrow = TRUE)
  zeta.max[j] &lt;- Ulist[which.max(rowSums(zeta))]
}

table(zeta.max)</code></pre>
<pre><code>zeta.max
Emp_cov    PC_1    PC_2    PC_3    PC_4    PC_5    tPCA 
      1      16      12       2      29       5      95 </code></pre>
<pre class="r"><code>### plot the top eigenvector capturing the predominant effect-sharing patterns
cols.trt &lt;- readRDS(&quot;output/Ly6C-/condition_colors.Rds&quot;)
trts &lt;- names(cols.trt)
R &lt;- length(trts)

eig.tPCA &lt;- eigen(fit$Ulist[[1]])
pve.tPCA &lt;- eig.tPCA$values/sum(eig.tPCA$values)

v &lt;- eig.tPCA$vectors[,1]
  
barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.4, las = 2, col = cols.trt,
        main = paste0(names(fit$Ulist)[1], &quot;: weight &quot;, round(fit$pi[1],3), &quot;:\nEigenvector &quot;, 1, &quot; (pve = &quot;, round(pve.tPCA[1],3), &quot;)&quot;))</code></pre>
<p><img src="figure/sim_null_Ly6C.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### identified genes for which the data-driven tPCA has the largest weight 
gene.list &lt;- genes[zeta.max==&quot;tPCA&quot;]

### for this subset of genes, compute the gene-wise correlation across all cells
data.sub &lt;- as.matrix(data[gene.list,])
corr.sub &lt;- cor(t(data.sub))

pheatmap(corr.sub, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=4, main=&quot;Gene-wise correlation for the identified genes \n (largest weight: tPCA)&quot;)</code></pre>
<p><img src="figure/sim_null_Ly6C.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### for a randomly chosen subset of genes of the same size, compute the gene-wise correlation across all cells
set.seed(100)
data.tmp &lt;- as.matrix(data[sample(1:nrow(data), length(gene.list), replace=FALSE),])
corr.tmp &lt;- cor(t(as.matrix(data.tmp)))

pheatmap(corr.tmp, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=4, main=&quot;Gene-wise correlation for a randomly chosen set of genes&quot;)</code></pre>
<p><img src="figure/sim_null_Ly6C.Rmd/unnamed-chunk-3-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="explore-why-such-strong-correlation-can-happen-for-genes-in-single-cell-data" class="section level2">
<h2>Explore why such strong correlation can happen for genes in single cell data</h2>
<pre class="r"><code>### perform hierarchical clustering of this subset of genes based on pairwise correlation
hcluster &lt;- hclust(as.dist(1-corr.sub), method=&quot;complete&quot;)
plot(hcluster, cex=0.5, xlab=&quot;clustering based on correlation matrix&quot;)</code></pre>
<p><img src="figure/sim_null_Ly6C.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### look at the gene names
rownames(corr.sub)[hcluster$order]</code></pre>
<pre><code> [1] &quot;Rcsd1&quot;    &quot;Snx5&quot;     &quot;S100a10&quot;  &quot;Rps15&quot;    &quot;Eef1b2&quot;   &quot;Rps7&quot;    
 [7] &quot;Rps26&quot;    &quot;Rps6&quot;     &quot;Rpl5&quot;     &quot;Eef2&quot;     &quot;Rps18&quot;    &quot;Rpl22&quot;   
[13] &quot;Rpl7&quot;     &quot;Rpl10a&quot;   &quot;Eef1a1&quot;   &quot;Rpsa&quot;     &quot;Rpl28&quot;    &quot;Rpl27a&quot;  
[19] &quot;Rpl17&quot;    &quot;Rps9&quot;     &quot;Tpt1&quot;     &quot;Rpl3&quot;     &quot;Rps8&quot;     &quot;Rps20&quot;   
[25] &quot;Rps13&quot;    &quot;Rps15a&quot;   &quot;Rpl32&quot;    &quot;Rpl23&quot;    &quot;Rps10&quot;    &quot;Rps16&quot;   
[31] &quot;Rps3a1&quot;   &quot;Rps3&quot;     &quot;Rpl8&quot;     &quot;Rpl18a&quot;   &quot;Rps24&quot;    &quot;Rpl19&quot;   
[37] &quot;Rps5&quot;     &quot;Rpl13&quot;    &quot;Rplp1&quot;    &quot;Rpl12&quot;    &quot;Rps19&quot;    &quot;Rpl13a&quot;  
[43] &quot;Shisa5&quot;   &quot;mt-Nd2&quot;   &quot;mt-Nd4&quot;   &quot;mt-Nd1&quot;   &quot;mt-Nd4l&quot;  &quot;mt-Cytb&quot; 
[49] &quot;mt-Co1&quot;   &quot;mt-Co2&quot;   &quot;mt-Atp6&quot;  &quot;mt-Co3&quot;   &quot;Tbc1d10c&quot; &quot;H2-DMb2&quot; 
[55] &quot;Fcmr&quot;     &quot;H2-Ob&quot;    &quot;Hfe&quot;      &quot;Gda&quot;      &quot;Srgn&quot;     &quot;Fxyd5&quot;   
[61] &quot;Ifitm2&quot;   &quot;Ndufb5&quot;   &quot;Zyx&quot;      &quot;Mpeg1&quot;    &quot;Sept9&quot;    &quot;Ifitm6&quot;  
[67] &quot;Itgal&quot;    &quot;Tln1&quot;     &quot;Msrb1&quot;    &quot;Cst3&quot;     &quot;Gngt2&quot;    &quot;Sat1&quot;    
[73] &quot;Rap1a&quot;    &quot;Ctsb&quot;     &quot;Fyb&quot;      &quot;Hck&quot;      &quot;Cd68&quot;     &quot;Anp32a&quot;  
[79] &quot;Zfp706&quot;   &quot;Ostf1&quot;    &quot;H2-D1&quot;    &quot;Arpc2&quot;    &quot;Cyba&quot;     &quot;Arpc1b&quot;  
[85] &quot;Tmsb4x&quot;   &quot;Tyrobp&quot;   &quot;Fcer1g&quot;   &quot;Ifitm3&quot;   &quot;Gng5&quot;     &quot;Pfn1&quot;    
[91] &quot;Calm1&quot;    &quot;H3f3b&quot;    &quot;Myl12b&quot;   &quot;Rgs2&quot;     &quot;Il10ra&quot;  </code></pre>
<pre class="r"><code>### identify a subset of genes that are strongly correlated with each other
idx.cluster &lt;- hcluster$order[12:38]
corr.cluster &lt;- corr.sub[idx.cluster, idx.cluster]
pheatmap(corr.cluster, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=6, fontsize_col=6, 
         main=&quot;Gene-wise correlation for this subset of 27 genes&quot;)</code></pre>
<p><img src="figure/sim_null_Ly6C.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="findings" class="section level2">
<h2>Findings</h2>
<p>1. The <a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/master/output/Ly6C-/sim_null/scatterplots.pdf">scatter plots</a>, which plot the raw single cell count data for every pair of genes in this subset, suggest that the single cell data are strongly correlated for this subset of genes.</p>
<p>2. The <a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/master/output/Ly6C-/sim_null/scatterplots_norm.pdf">scatter plots</a>, which plot the normalized single cell data for every pair of genes in this subset, suggest that the normalized data fed into mash are much less correlated for this subset of genes.</p>
<p>3. The <a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/master/output/Ly6C-/sim_null/plots.pdf">scatter plots</a>, which compare the observed and expected counts for each condition in this null dataset, suggest consistent effect-sharing patterns across this subset of genes.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gridExtra_2.3   pheatmap_1.0.12 Matrix_1.2-18   workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4.6       knitr_1.30         whisker_0.4        magrittr_1.5      
 [5] munsell_0.5.0      colorspace_1.4-1   lattice_0.20-41    R6_2.4.1          
 [9] rlang_0.4.7        stringr_1.4.0      tools_3.6.2        grid_3.6.2        
[13] gtable_0.3.0       xfun_0.17          git2r_0.27.1       htmltools_0.4.0   
[17] yaml_2.2.1         digest_0.6.25      rprojroot_1.3-2    lifecycle_0.2.0   
[21] RColorBrewer_1.1-2 later_1.0.0        promises_1.1.0     fs_1.4.1          
[25] glue_1.4.1         evaluate_0.14      rmarkdown_2.3      stringi_1.4.6     
[29] compiler_3.6.2     scales_1.1.1       backports_1.1.7    httpuv_1.5.2      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
