<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Application of poisson mash and mash to simulated single cell data (setting 3)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">poisson_mash_random_effects</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/poisson_mash_random_effects">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Application of poisson mash and mash to simulated single cell data (setting 3)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2020-11-1</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-11-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>poisson_mash_random_effects/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201102code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201102)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201102code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201102)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstree3cd5c87e34b595b91f962e3894c2c46ca224da18targetblank3cd5c87a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/3cd5c87e34b595b91f962e3894c2c46ca224da18" target="_blank">3cd5c87</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashrandomeffectstree3cd5c87e34b595b91f962e3894c2c46ca224da18targetblank3cd5c87a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/poisson_mash_random_effects/tree/3cd5c87e34b595b91f962e3894c2c46ca224da18" target="_blank">3cd5c87</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/Ly6C-/.DS_Store

Untracked files:
    Untracked:  output/B_cells/
    Untracked:  output/Ly6C-/condition_colors.Rds
    Untracked:  output/Ly6C-/sim_null/ed_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_fit_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/pois_mash_posterior_rep_3.Rds
    Untracked:  output/Ly6C-/sim_null/raw_data3.Rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/sim_setting3_B_cells.Rmd</code>) and HTML (<code>docs/sim_setting3_B_cells.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/poisson_mash_random_effects/blob/3cd5c87e34b595b91f962e3894c2c46ca224da18/analysis/sim_setting3_B_cells.Rmd" target="_blank">3cd5c87</a>
</td>
<td>
yushaliu
</td>
<td>
2020-11-02
</td>
<td>
Add analysis results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-procedure" class="section level2">
<h2>Simulation procedure</h2>
<p>To simulate single cell data, we start with a subset of the single cell cytokines data, which include 1) 2,096 cells from <span class="math inline">\(R=25\)</span> conditions from the second batch, and 2) 5,081 genes that have at least 100 total counts over these 25 conditions. We remove genes that have relatively few reads because we will add treatment and/or individual sample effects to the data via thinning.</p>
<div id="individual-sample-effect" class="section level3">
<h3>Individual sample effect</h3>
<p>Starting with the real dataset, we randomly permute the cells with respect to the condition labels to create null data. Then we use thinning to add individual sample effect <span class="math inline">\(\eta_j \sim MVN(0, \: \psi_j^2 I_R)\)</span> to all genes. We simulate <span class="math inline">\(\psi_j\)</span> from the set <span class="math inline">\(\log_2(\exp(1))*\{0.001, 0.05, 0.1, 0.2\}\)</span> with probablity <span class="math inline">\(\{0.3, 0.4, 0.2, 0.1\}\)</span>, which resembles the empirical distribution of <span class="math inline">\(\psi_j\)</span> we see from the poisson mash fit to the real dataset.</p>
</div>
<div id="treatment-effect" class="section level3">
<h3>Treatment effect</h3>
<p>We also use thinning to add treatment effect <span class="math inline">\(\beta_j \sim \sum_{k=1}^3 \pi_k MVN(0, w_j^2 U_k)\)</span>. The prior covariances <span class="math inline">\(U_k\)</span>’s are learned from the poisson mash fit to the original dataset and are given as follows (all <span class="math inline">\(U_k\)</span>’s are rank-1 matrices and can be expressed as <span class="math inline">\(U_k = u_k u_k&#39;\)</span>), with the prior weights <span class="math inline">\(\pi_1 = 0.5, \pi_2=0.25, \pi_3=0.25\)</span>. We simulate the effect size <span class="math inline">\(w_j^2\)</span> to be realistic and to yield reasonable ROC curves (as shown later).</p>
<pre class="r"><code># conditions and colors
cols.trt &lt;- readRDS(&quot;output/B_cells/condition_colors.Rds&quot;)
trts &lt;- names(cols.trt)
ulist.sim &lt;- readRDS(&quot;output/B_cells/ulist_sim.rds&quot;)

for (k in 1:3){
  v &lt;- ulist.sim$ulist[[k]]
  
  barplot(v/v[which.max(abs(v))], names = names(cols.trt), cex.names = 0.75, col = cols.trt, las = 2,
          main = paste0(&quot;u&quot;, k, &quot;: weight &quot;, ulist.sim$pi.u[k]))
}</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-1-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>We consider the following three scenarios regarding the addition of treatment effects. For each scenario, we simulate 5 single cell datasets.</p>
<p>Scenario 1. No treatment effect is added.</p>
<p>Scenario 2. Treatment effect is added to <span class="math inline">\(500\)</span> genes randomly chosen from the genes with relatively high expression (&gt; 70 percentile). The effect size <span class="math inline">\(w_j\)</span> is drawn from the set <span class="math inline">\(\{\log_2(1.5), \log_2(2), \log_2(2.5), \log_2(3)\}\)</span> with probability <span class="math inline">\(\{0.2, 0.5, 0.2, 0.1\}\)</span>, and further added an <span class="math inline">\(N(0, 0.1^2)\)</span> noise on top of that.</p>
<p>Scenario 3. Treatment effect is added to <span class="math inline">\(500\)</span> genes randomly chosen from the genes with relatively lower expression levels (35 ~ 70 percentile). The effect size <span class="math inline">\(w_j\)</span> is drawn from the set <span class="math inline">\(\{\log_2(2), \log_2(3), \log_2(4), \log_2(5)\}\)</span> with probability <span class="math inline">\(\{0.3, 0.4, 0.2, 0.1\}\)</span>, and further added an <span class="math inline">\(N(0, 0.1^2)\)</span> noise on top of that.</p>
</div>
</div>
<div id="scenario-1-no-treatment-effect" class="section level2">
<h2>Scenario 1 (no treatment effect)</h2>
<pre class="r"><code>res &lt;- readRDS(&quot;output/B_cells/simulations3/null_random_effects_summary.Rds&quot;)

### run time of poisson with random effects
res$runtime.pois.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
3212.037    4.621 3218.601 

[[2]]
    user   system  elapsed 
3831.990    7.013 3841.503 

[[3]]
    user   system  elapsed 
3493.764    5.361 3501.456 

[[4]]
    user   system  elapsed 
3258.664    4.247 3264.267 

[[5]]
    user   system  elapsed 
3264.845    4.797 3271.174 </code></pre>
<pre class="r"><code>### run time of mash with random effects
res$runtime.mash.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
7637.886   15.247 7656.467 

[[2]]
    user   system  elapsed 
6646.552   11.405 6660.712 

[[3]]
    user   system  elapsed 
7430.417   13.003 7445.890 

[[4]]
    user   system  elapsed 
7401.189   13.081 7416.433 

[[5]]
    user   system  elapsed 
6764.708   12.070 6778.546 </code></pre>
<pre class="r"><code>### false positives found by poisson mash with random effects
res$num.effects.pois</code></pre>
<pre><code>[1]  9 34 20 19 13</code></pre>
<pre class="r"><code>### false positives found by mash with random effects
res$num.effects.mash</code></pre>
<pre><code>[1] 15 11 15 22 17</code></pre>
</div>
<div id="scenario-2-treatment-effect-added-to-highly-expressed-genes" class="section level2">
<h2>Scenario 2 (treatment effect added to highly expressed genes)</h2>
<pre class="r"><code>res &lt;- readRDS(&quot;output/B_cells/simulations3/sim_signal_upr_summary.Rds&quot;)
alpha &lt;- res$alpha

### run time of poisson with random effects
res$runtime.pois.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
6372.429   10.051 6386.492 

[[2]]
    user   system  elapsed 
6228.360    9.449 6241.152 

[[3]]
    user   system  elapsed 
6567.847   14.530 6587.265 

[[4]]
    user   system  elapsed 
7332.019   15.512 7354.678 

[[5]]
    user   system  elapsed 
7440.947   17.482 7465.518 </code></pre>
<pre class="r"><code>### run time of mash with random effects
res$runtime.mash.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
8649.560   17.284 8670.450 

[[2]]
    user   system  elapsed 
9633.079   18.875 9655.764 

[[3]]
    user   system  elapsed 
9346.277   19.021 9370.374 

[[4]]
    user   system  elapsed 
8941.832   16.446 8962.445 

[[5]]
    user   system  elapsed 
9462.680   17.282 9484.129 </code></pre>
<pre class="r"><code>### look at the FDR for identifying units at alpha=0.05
res$fdr.pois[,alpha==0.05]</code></pre>
<pre><code>[1] 0.02230483 0.04000000 0.05019305 0.03149606 0.14028777</code></pre>
<pre class="r"><code>res$fdr.mash[,alpha==0.05]</code></pre>
<pre><code>[1] 0.04680851 0.11111111 0.06976744 0.07627119 0.06190476</code></pre>
<pre class="r"><code>### look at the FSR for identifying unit-condition pairs at alpha=0.05
res$fsr.pair.pois[,alpha==0.05]</code></pre>
<pre><code>[1] 0.05195291 0.04762777 0.05688337 0.04390716 0.13301172</code></pre>
<pre class="r"><code>res$fsr.pair.mash[,alpha==0.05]</code></pre>
<pre><code>[1] 0.06550549 0.11252512 0.12644536 0.08256536 0.08125844</code></pre>
<pre class="r"><code>### plot ROC for detecting non-null units 
par(mfrow=c(2,3))

for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;All units: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr1.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u1: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr1.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr2.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u2: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr2.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-4-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr3.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u3: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr3.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-4-4.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### plot ROC for detecting non-null unit-condition pairs 
par(mfrow=c(2,3))

for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;All pairs: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr1.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u1: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr1.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr2.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u2: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr2.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-5-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr3.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u3: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr3.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-5-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="scenario-3-treatment-effect-added-to-lowly-expressed-genes" class="section level2">
<h2>Scenario 3 (treatment effect added to lowly expressed genes)</h2>
<pre class="r"><code>res &lt;- readRDS(&quot;output/B_cells/simulations3/sim_signal_lwr_summary.Rds&quot;)
alpha &lt;- res$alpha

### run time of poisson with random effects
res$runtime.pois.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
5451.559    8.192 5461.926 

[[2]]
    user   system  elapsed 
5220.639    8.297 5231.030 

[[3]]
    user   system  elapsed 
4824.350    7.485 4833.695 

[[4]]
    user   system  elapsed 
4898.114    7.080 4907.118 

[[5]]
    user   system  elapsed 
5212.032    7.351 5221.534 </code></pre>
<pre class="r"><code>### run time of mash with random effects
res$runtime.mash.fit</code></pre>
<pre><code>[[1]]
    user   system  elapsed 
9357.905   16.152 9378.097 

[[2]]
    user   system  elapsed 
9851.027   18.412 9873.678 

[[3]]
    user   system  elapsed 
9251.888   17.087 9272.942 

[[4]]
    user   system  elapsed 
9014.131   17.533 9035.530 

[[5]]
    user   system  elapsed 
9196.779   16.431 9216.066 </code></pre>
<pre class="r"><code>### look at the FDR for identifying units at alpha=0.05
res$fdr.pois[,alpha==0.05]</code></pre>
<pre><code>[1] 0.08974359 0.11904762 0.08620690 0.07627119 0.12800000</code></pre>
<pre class="r"><code>res$fdr.mash[,alpha==0.05]</code></pre>
<pre><code>[1] 0.10218978 0.13475177 0.11594203 0.12000000 0.07317073</code></pre>
<pre class="r"><code>### look at the FSR for identifying unit-condition pairs at alpha=0.05
res$fsr.pair.pois[,alpha==0.05]</code></pre>
<pre><code>[1] 0.09044415 0.13609272 0.11246095 0.06285311 0.12726684</code></pre>
<pre class="r"><code>res$fsr.pair.mash[,alpha==0.05]</code></pre>
<pre><code>[1] 0.12594268 0.15785997 0.12073214 0.14174077 0.07596852</code></pre>
<pre class="r"><code>### plot ROC for detecting non-null units 
par(mfrow=c(2,3))

for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;All units: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr1.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u1: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr1.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-7-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr2.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u2: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr2.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-7-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.mash[iter,] , res$tpr3.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Units from u3: iter&quot;, iter))
  lines(res$fpr.pois[iter,], res$tpr3.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-7-4.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### plot ROC for detecting non-null unit-condition pairs 
par(mfrow=c(2,3))

for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;All pairs: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr1.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u1: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr1.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-8-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr2.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u2: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr2.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}

plot.new()</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-8-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>for (iter in 1:5){
  plot(res$fpr.pair.mash[iter,] , res$tpr3.pair.mash[iter,], type=&quot;l&quot;, xlim=c(0, 0.2), ylim=c(0, 1), 
       xlab=&quot;False positive rate&quot;, ylab=&quot;True positive rate&quot;, main=paste0(&quot;Pairs from u3: iter&quot;, iter))
  lines(res$fpr.pair.pois[iter,], res$tpr3.pair.pois[iter,], lty=4, col=&quot;magenta&quot;)
  if(iter %% 5 ==1)
  {
    legend(&quot;bottomright&quot;, legend=c(&quot;mash + re&quot;, &quot;pois mash + re&quot;), text.col=c(&quot;black&quot;, &quot;magenta&quot;), col=c(&quot;black&quot;, &quot;magenta&quot;), lty=c(1,4), bty=&quot;n&quot;)
  }
}</code></pre>
<p><img src="figure/sim_setting3_B_cells.Rmd/unnamed-chunk-8-4.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4.6    rprojroot_1.3-2 digest_0.6.25   later_1.0.0    
 [5] R6_2.4.1        backports_1.1.7 git2r_0.27.1    magrittr_1.5   
 [9] evaluate_0.14   stringi_1.4.6   rlang_0.4.7     fs_1.4.1       
[13] promises_1.1.0  whisker_0.4     rmarkdown_2.3   tools_3.6.2    
[17] stringr_1.4.0   glue_1.4.1      httpuv_1.5.2    xfun_0.17      
[21] yaml_2.2.1      compiler_3.6.2  htmltools_0.4.0 knitr_1.30     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
